### ISSUE TITLE
Should `Layout/ClassStructure` allow more general groupings?

### ISSUE THREAD

**Issue by marcandre** [rubocop/rubocop#9726]
Currently, one can define "categories", for example `{"attributes" => ["attr_reader", "attr_writer", "attr_accessor"]}` and put `"attributes"` somewhere in the class order.

This has the convenience of allowing arbitrary order within that category (e.g. a `attr_reader` can follow or preceede `attr_writer`)

Categories are limited to macros.

This setup has the limitation that "builtin" categories must have a strict order between them. For example, it is not possible to specify that instances methods and class methods should all be after constants and before private methods, but in any order between them.

Allowing the config to have arrays, e.g. `"ExpectedOrder" => ["constants", ["public_methods", "public_class_methods"], "private_methods"]` could allow this. It would make the "Categories" kind of redundant too.

Is this a worthwhile feature?

Note: current cop handles class methods really badly, but upcoming refactor will be able to distinguish private class methods, private constants, etc.

**Comment by marcandre** [comment]
Alternatively, we could allow "buitin" and categories in categories:
```
"Categories" => {"public" => ["public_methods", "public_class_methods"]}
"ExpectedOrder" => ["constants", "public", "private_methods"]
```

**Comment by bbatsov** [comment]
I think we went with the built-in categories with the intention to have a relatively simple configuration that still allowed some degree of tuning. Both ideas you have make sense to me, although I'm a bit concerned we might end up a configuration that'd be harder for the users to understand. 

Btw, by "built-in" you mean the categories defined by the default configuration, right?

**Comment by marcandre** [comment]
By builtin I mean the categories that we hardcode, i.e. "methods", "class_methods", "initializer", "constants".

**Comment by marcandre** [comment]
For example, after my refactor, we can finally distinguish private constants and private class methods.

I would like the private constants to be moved after the public stuff, but I'm not sure there's a good reason for them to be strictly before or after private methods (instance or class), for example. 

Maybe an ideal config would be:

```yaml
  Categories:
    module_inclusion:
      - include
      - prepend
      - extend
  ExpectedOrder:
      - module_inclusion
      - public_constants
      - public_class_methods
      - initializer
      - public_methods
      - protected_methods
      - - private_constants
        - - private_class_methods
          - private_methods
```

Which would mean "after the protected methods, put the private stuff in any order, except that class methods should come before private methods.
May be over-complicated though. What do you think the best class structure should look like?
